programs/lib/syscalls/syscalls.c

```diff
@@ -175,3 +176,21 @@ int _gettimeofday(struct timeval *tv, void *tz) {
     tv->tv_usec = ns_to_us(ns % (int) (1e9));
     return 0;
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// Malloc required functions
+////////////////////////////////////////////////////////////////////////////////
+
+static fp_lock_t malloc_lock = FP_LOCK_INITIALIZER;
+
+FP_TEST_OVERRIDE
+void __malloc_lock(struct _reent *r)
+{
+    fp_lock_acquire(&malloc_lock);
+}
+
+FP_TEST_OVERRIDE
+void __malloc_unlock(struct _reent *r)
+{
+    fp_lock_release(&malloc_lock);
+}
```

Comment:
Newlib's malloc requires the user to implement these functions to work in a multi-threaded environment. The `FP_TEST_OVERRIDE` is used to check that malloc fails if these are implemented as doing nothing in a test.